src/Runtime/PyExecutionSession.cpp

```diff
@@ -19,74 +19,72 @@ namespace onnx_mlir {
   }
 
   std::vector<py::array> outputPyArrays;
   auto *wrappedOutput = _entryPointFunc(wrappedInput);
-  for (int i = 0; i < numRtMemRefs(wrappedOutput); i++) {
-    auto *dynMemRef = getRtMemRef(wrappedOutput, i);
+  for (int i = 0; i < rmrListGetNumRmrs(wrappedOutput); i++) {
+    auto *rmr = rmrListGetRmrByIndex(wrappedOutput, i);
     auto shape = std::vector<int64_t>(
-        dynMemRef->sizes, dynMemRef->sizes + dynMemRef->rank);
+        rmrGetDataShape(rmr), rmrGetDataShape(rmr) + rmrGetRank(rmr));
 
     // https://numpy.org/devdocs/user/basics.types.html
     py::dtype dtype;
-    if (dynMemRef->onnx_dtype == onnx::TensorProto::FLOAT)
+    if (rmrGetDataType(rmr) == onnx::TensorProto::FLOAT)
       dtype = py::dtype("float32");
-    else if (dynMemRef->onnx_dtype = onnx::TensorProto::UINT8)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::UINT8)
       dtype = py::dtype("uint8");
-    else if (dynMemRef->onnx_dtype = onnx::TensorProto::INT8)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::INT8)
       dtype = py::dtype("int8");
-    else if (dynMemRef->onnx_dtype = onnx::TensorProto::UINT16)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::UINT16)
       dtype = py::dtype("uint16");
-    else if (dynMemRef->onnx_dtype = onnx::TensorProto::INT16)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::INT16)
       dtype = py::dtype("int16");
-    else if (dynMemRef->onnx_dtype == onnx::TensorProto::INT32)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::INT32)
       dtype = py::dtype("int32");
-    else if (dynMemRef->onnx_dtype == onnx::TensorProto::INT64)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::INT64)
       dtype = py::dtype("int64");
     // TODO(tjingrant) wait for Tong's input for how to represent string.
-    else if (dynMemRef->onnx_dtype = onnx::TensorProto::BOOL)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::BOOL)
       dtype = py::dtype("bool_");
-    else if (dynMemRef->onnx_dtype = onnx::TensorProto::FLOAT16)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::FLOAT16)
       dtype = py::dtype("float32");
-    else if (dynMemRef->onnx_dtype = onnx::TensorProto::DOUBLE)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::DOUBLE)
       dtype = py::dtype("float64");
-    else if (dynMemRef->onnx_dtype == onnx::TensorProto::UINT32)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::UINT32)
       dtype = py::dtype("uint32");
-    else if (dynMemRef->onnx_dtype == onnx::TensorProto::UINT64)
+    else if (rmrGetDataType(rmr) == onnx::TensorProto::UINT64)
       dtype = py::dtype("uint64");
     else {
-      fprintf(stderr, "Unsupported ONNX type in RtMemRef.onnx_dtype.");
+      fprintf(stderr, "Unsupported ONNX type in RtMemRef.");
       exit(1);
     }
 
-    outputPyArrays.emplace_back(py::array(dtype, shape, dynMemRef->data));
+    outputPyArrays.emplace_back(py::array(dtype, shape, rmrGetData(rmr)));
   }
 
```

Comment:
It's a bit of a technical reason if I recall correctly. Otherwise, we need to create another struct where the pointer `void* data` is a const pointer, etc...

Reply:
So this makes destroying `wrappedInput` a bit tricky. When data is writable, `rmrListDestroy(wrappedInput)` should free the rmr structs and the rmrList struct **without** freeing data. Since the data belongs to the caller of `pyRun`. When data is not writable, data should be freed since we malloc-ed and copied it.

So I think when data is writable, we should create `inputRtMemRef` with `rmrCreateWithNameAndOwnership()` and set ownership to false. This means that `inputRtMemRef` does not own the data. If we do that, `rmrListDestroy(wrappedInput)` should do the right thing for both cases.

Reply:
That sounds reasonable! Will update the code as you suggested.

Reply:
Thanks. That's all I got for now. JNI code needs minor update but I will do that with a separate patch once this gets merged.